### PR TITLE
support bindValue with typeahead

### PR DESCRIPTION
**Pull request by anjmao**
fixes https://github.com/ng-select/ng-select/issues/282

### CODE REVIEW

File: src/ng-select/ng-select.component.ts [hunk 1]
```diff
@@ -489,14 +489,19 @@ export class NgSelectComponent implements OnDestroy, OnChanges, AfterViewInit, C
...
-            const item = this.itemsList.findItem(val);
+            let item = this.itemsList.findItem(val);
             if (item) {
                 this.itemsList.select(item);
             } else {
                 const isObject = val instanceof Object;
                 const simpleValue = !isObject && !this.bindValue;
                 if (isObject || simpleValue) {
                     this.itemsList.select(this.itemsList.mapItem(val, simpleValue, null));
+                } else if (!isObject && this.bindValue && this._isTypeahead) {
```
Comment (varnastadeus) [inline src/ng-select/ng-select.component.ts]: https://github.com/ng-select/ng-select/pull/312/files#diff-504437db9e7ef93a33fdb3cad468d194R418
what about this line? is it really necessary?

Comment (varnastadeus) [reply src/ng-select/ng-select.component.ts]: This won't work. Imagine you have id as `bindValue`, everytime you look up new items, labels will switch to id's 👹

Comment (anjmao) [reply src/ng-select/ng-select.component.ts]: That is correct, but in this case when typeahead is used with bindValue you don't have any other choose but only show bindValue as label, so user is responsible for it and he can show something that make sense, city name, user name, email ....

Comment (varnastadeus) [reply src/ng-select/ng-select.component.ts]: But it shouldn't happen for correctly selected items. When you select you have normal label, and after lookup it switches to id..

Comment (anjmao) [reply src/ng-select/ng-select.component.ts]: True, solution by not removing selected items sounds better 👍

File: src/ng-select/ng-select.component.ts [hunk 2]
```diff
@@ -486,20 +487,27 @@ export class NgSelectComponent implements OnDestroy, OnChanges, AfterViewInit, C
...
             } else {
-                const isObject = val instanceof Object;
-                const simpleValue = !isObject && !this.bindValue;
-                if (isObject || simpleValue) {
+                const isValObject = isObject(val)
+                const isPrimitive = !isValObject && !this.bindValue;
+                if ((isValObject || isPrimitive)) {
                     this.itemsList.select(this.itemsList.mapItem(val, null));
+                } else if (this.bindValue && !this._isTypeahead) {
+                    item = {};
```
Comment (varnastadeus) [inline src/ng-select/ng-select.component.ts]:
```javascript 
item = {
    [this.bindLabel]: null,
    [this.bindValue]: val,
};
```

File: src/ng-select/ng-select.component.ts [hunk 3]
```diff
@@ -511,11 +519,24 @@ export class NgSelectComponent implements OnDestroy, OnChanges, AfterViewInit, C
         } else {
             select(ngModel);
         }
+
+        this._updateSelectedItems();
+    }
+
+    private _updateSelectedItems() {
+        this.selectedItems.forEach((s, i) => {
```
Comment (varnastadeus) [inline src/ng-select/ng-select.component.ts]: rename `s` to `item`?

File: src/ng-select/ng-select.component.ts [hunk 4]
```diff
@@ -486,20 +487,27 @@ export class NgSelectComponent implements OnDestroy, OnChanges, AfterViewInit, C
...
+    private _handleWriteValue(ngModel: any | any[]) {
+        const isEmptyArray = ngModel && Array.isArray(ngModel) && ngModel.length === 0;
+        if (ngModel === null || isEmptyArray) {
+            this.itemsList.clearSelected();
             return;
         }
 
         const select = (val: any) => {
-            const item = this.itemsList.findItem(val);
+            let item = this.itemsList.findItem(val);
```
Comment (varnastadeus) [inline src/ng-select/ng-select.component.ts]: `_handleWriteValue` and `_updateSelectedItems` are searching for item 2 times even if former finds it. I would suggest removing `let item = this.itemsList.findItem(val);` in the first method or marking it as `mapped` or something so `_updateSelectedItems` would search only unmapped items?

File: src/ng-select/ng-select.component.ts (hunk 2 above)
Comment (varnastadeus) [reply src/ng-select/ng-select.component.ts]: Also I think we should support this use case for typeahead too. Users would need to preload items to see selected.
```javascript
else if (this.bindValue) {
```

Comment (varnastadeus) [reply src/ng-select/ng-select.component.ts]: screen shot 2018-03-07 at 16 18 08
on multiselect, label is visible

**Pull request closed without merging**